Support multiple crate versions in --extern-html-root-url

kornelski opened the pull request:
Rustdoc's `--extern-html-root-url` used to use `tcx.crate_name()` to identify crates, but that used crates' internal names from their metadata, instead of names given to them in `--extern`. That was confusing, because both `--extern…` arguments seem related and use similar syntax. Crucially, this didn't work correctly with Cargo's package aliases or multiple versions of crates.

`sess.opts.externs` lacks `CrateNum`, and `Resolver.extern_prelude` gets destroyed before `rustdoc` has a chance to see it, so I've had to save this mapping in `CStore`.

Just in case, I've kept the previous mapping by crate name as a fallback for crates that weren't matched by their extern name.

Fixes rust-lang/rust#76296

kornelski committed:
Clarify update_extern_crate

kornelski committed:
Support multiple crate versions in --extern-html-root-url

kornelski committed:
Test renamed crates in rustdoc

kornelski commented:
And f9f6a70e3331f5792b473cc214aeee9690f50101 is just a bit of redundant code I've noticed in the process, too small to give it a separate PR.

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16083135284/job/45390679190) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45390679190) [(plain)](https://github.com/rust-lang/rust/commit/63deae4992965e398e6245006bdf346eb30bd568/checks/45390679190/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
Diff in /checkout/src/librustdoc/formats/cache.rs:180:
 
             let name = e.name(tcx);
             cx.cache.extern_locations.entry(e.crate_num).or_insert_with(|| {
-                let extern_url = render_options.extern_html_root_urls.get(name.as_str()).map(|u| &**u);
+                let extern_url =
+                    render_options.extern_html_root_urls.get(name.as_str()).map(|u| &**u);
                 e.location(extern_url, extern_url_takes_precedence, dst, tcx)
             });
             cx.cache.external_paths.insert(e.def_id(), (vec![name], ItemType::Module));
fmt: checked 6143 files
Build completed unsuccessfully in 0:00:46
  local time: Sat Jul  5 01:28:34 UTC 2025
  network time: Sat, 05 Jul 2025 01:28:34 GMT
##[error]Process completed with exit code 1.
```

</details>

jyn514 commented:
cc @aDotInTheVoid @obi1kenobi - i haven’t checked if this will help distinguish versions for the json backend too but i think it might

kornelski commented:
Yes, the URL is in the JSON, and can be creatively used to identify the crates.

jyn514 commented:
to be clear i do *not* think html-root-url should be used for this, i’m saying the new metadata you’ve added to CrateRoot might allow the json backend to emit structured metadata.

lolbinarycat commented:
I think the name resolution method used here should be documented [in the rustdoc book](https://doc.rust-lang.org/nightly/rustdoc/unstable-features.html#--extern-html-root-url-control-how-rustdoc-links-to-non-local-crates)

lolbinarycat commented on compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs:
```diff
@@ -630,14 +630,29 @@ impl CStore {
+        name: Symbol,
+        extern_crate: ExternCrate,
+    ) {
+        debug_assert_eq!(extern_crate.dependency_of, LOCAL_CRATE);
```
```suggestion
        // this function should not be called on transitive dependencies
        debug_assert_eq!(extern_crate.dependency_of, LOCAL_CRATE);
```

lolbinarycat commented on src/librustdoc/formats/cache.rs:
```diff
@@ -158,18 +159,31 @@ impl Cache {
+        }
+
         // Cache where all our extern crates are located
         // FIXME: this part is specific to HTML so it'd be nice to remove it from the common code
```
if the doc url is emitted in the json, doesn't that make this fixme comment wrong? if so, i would replace it with a note saying that this needs to be common because it is actually emitted by both backends.

kornelski replied:
I've added a comment

lolbinarycat commented on tests/rustdoc/extern/extern-html-alias.rs:
```diff
@@ -0,0 +1,8 @@
+//@ compile-flags:-Z unstable-options --extern-html-root-url extern_prelude_name=https://renamed.example.com  --extern-html-root-url empty=https://bad.invalid
+//@ aux-crate:extern_prelude_name=empty.rs
+
+extern crate extern_prelude_name as internal_ident_name;
+
+//@ has extern_html_alias/index.html
+//@ has - '//a/@href' 'https://renamed.example.com/empty/index.html'
+pub use internal_ident_name as yet_different_name;
```
1. `internal_ident_name` is a misleading name, as it actually does exist in extern prelude as an alias (i.e. `::internal_ident_name` would be valid here, in this case both exist in the extern prelude)
2. we should test that `extern crate as` aliases are handled correctly
3. we should test the fallback behavior (using crate name)

Doing all this would probably require at least a second `aux-crate`.

kornelski replied:
`Resolver.extern_prelude` confused me, because it contains only `core`, `std`, and `extern_prelude_name`, and not `internal_ident_name`. 

I've renamed things to avoid referring to the extern prelude.

kornelski commented on src/doc/rustdoc/src/unstable-features.md:
```diff
@@ -395,6 +395,11 @@ flags to control that behavior. When the `--extern-html-root-url` flag is given
 in the output directory, those local docs will still override this flag.
 
+Crate names in this flag refer to the names in the [extern prelude], and support crates renamed
+via the `--extern` `rustc` flag.
```
This implementation intentionally doesn't support `as renamed` of `extern crate resolved as renamed`. This is because the renamed import is visible only in the source code, and not known the external tooling calling `rustdoc`.

I've changed the docs to avoid referring to the *extern prelude*.

aDotInTheVoid commented:
CC https://github.com/rust-lang/cargo/issues/8296

petrochenkov commented on compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs:
```diff
@@ -630,14 +630,32 @@ impl CStore {
+            extern_crate.dependency_of, LOCAL_CRATE,
+            "this function should not be called on transitive dependencies"
+        );
+        self.set_resolved_extern_crate_name(name, cnum);
```
I think it would be fine to do `if extern_crate.dependency_of == LOCAL_CRATE { set_resolved_extern_crate_name }` instead of splitting `update_extern_crate` into two function.
I doubt it will cause any perf regressions, but we can run the benchmarks.

kornelski replied:
I've done this for clarity, not performance. Updating local crate imports and updating transitive dependencies are two different tasks.

Logically, the caller should call both:

```
store.set_resolved_extern_crate_name(name);
store.update_extern_crate(deps);
```

but it would be a bit fragile, since the caller could do one thing but forget the other, which is why I have one entrypoint for the local crate, but the transitive deps update separately.

kornelski replied:
@petrochenkov I've split it into 117447f7e8aa01c09a234aaa74265cdfec3d5777 and dfdf5b80e48cb51d24014fff00ab33d6cf5c6758 to demonstrate the difference

petrochenkov commented:
This is better than I originally assumed, the crate lookup algorithm is not changed, only some already calculated data is exposed.
Also, delegating `--extern-html-root-url` to `--extern` seems like the best solution that we can do here.

kornelski committed:
Save names of used extern crates

Tracks association between `self.sess.opts.externs` (aliases in `--extern alias=rlib`) and resolved `CrateNum`

Intended to allow Rustdoc match the aliases in `--extern-html-root-url`

Force-injected extern crates aren't included, since they're meant for the linker only

kornelski committed:
Avoid redundant lookup in CrateLoader::existing_match

rust-log-analyzer commented:
The job **`x86_64-gnu-miri`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16454962219/job/46509634406) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46509634406) [(plain)](https://github.com/rust-lang/rust/commit/75c2eef61ee8d64da97b56b78c4a9f226d640fe7/checks/46509634406/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>
